accepted/install-locations.md

```diff
@@ -0,0 +1,227 @@
+* :new: per-user registered install location, in the `/host/fxr/<version>` subdirectory
+  * Windows - `HKCU\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation`
+  * Linux/macOS - `$HOME/.dotnet/install_location`
+* :new: global registered install location, in the `/host/fxr/<version>` subdirectory  
+  **We've committed to support this on Windows for .NET Core 3.0**
+  * Windows - `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` (32-bit view)  
+  * Linux/macOS - `/etc/dotnet/install_location`  
+* default global location, in the `/host/fxr/<version>` subdirectory
+  * Windows (32-bit process on 64-bit Windows) - `%ProgramFiles(x86)%\dotnet`
+  * Windows (everything else) - `%ProgramFiles%\dotnet`
+  * Linux - `/usr/share/dotnet`
+  * macOS - `/usr/local/share/dotnet`
+
+In each case with the `<version>` subdirectory, the latest available version as per SemVer 2.0 rules will be selected. Note that the algorithm picks the first location which exists, and only then looks for the `/host/fxr/<version>` subdirectory, and if that doesn't exist, or the library is missing it will fail.
+
+### Framework and SDK search
+All hosts use the same logic.
+
+Both framework and SDK search uses the same logic, the only difference is that frameworks are in the `share` subdirectory, while SDKs are in the `sdk` subdirectory of the install location.
+In both cases all locations listed below are searched and the complete list of all available SDKs/Frameworks is combined before making a choice which one will be used (based on version requirements).
+
+Search locations:
+* the location of the `hostfxr` used. See the `hostfxr` search algorithm above. This means that `DOTNET_ROOT` has effect on framework and SDK search as well.
+* If multi-level lookup is enabled (by default it is, can be disabled via `DOTNET_MULTILEVEL_LOOKUP=0` environment variable)
+  * :new: per-user registered install location
+    * Windows - `HKCU\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation`
+    * Linux/macOS - `$HOME/.dotnet/install_location`
+  * :new: global registered install location  
+    **We've committed to support this on Windows for .NET Core 3.0**
+    * Windows - `HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\<arch>\InstallLocation` (32-bit view)  
+    * Linux/macOS - `/etc/dotnet/install_location`
+  * default global location, in the `/host/fxr/<version>` subdirectory
+    * Windows (32-bit process on 64-bit Windows) - `%ProgramFiles(x86)%\dotnet`
+    * Windows (everything else) - `%ProgramFiles%\dotnet`
+    * :new: Linux - `/usr/share/dotnet`
+    * :new: macOS - `/usr/local/share/dotnet`
+
+*Note: The above means that on Linux/macOS in 2.2 the multi-level lookup was effectively non-functional, as there were no search paths for it. Part of the 3.0 changes is to change that and make Linux/macOS work similarly to Windows in this case as well.*
```

Comment:
We have been talking about phasing-out multi-level lookup, e.g.: https://github.com/dotnet/core-setup/issues/3606#issuecomment-373458531 . Have we changed our opinion?

Reply:
That issue is more of an SDK issue - if the user would invoke the app via `dotnet exec` it would behave as expected. The problem is with `dotnet run` which first finds SDK and then runs the app in the "context" of the SDK (using the same `dotnet.exe` and frameworks).

I think we will have some form of multi-level lookup going forward. If it's an environment variable or not is a separate discussion I think. I'll mention that issue in the doc.

If we end up not doing anything in this area in .NET Core 3.0 we're left where we're today which is very confusing. On Windows it works, on Linux/macOS it doesn't, for no good reason. The code mentions that "we don't have a global location" on Linux, but first that's not true already (since we do look into global location for shared frameworks and SDKs) and second we want the globally registered location for Linux as it's required by distros which will install to custom locations.

Reply:
> we will have some form of multi-level lookup going forward.

Multilevel-lookup stitching multiple global locations is fine, I think.

I think the problematic part of the multilevel-lookup is that it stitches your local unzipped SDK/runtime with the globally installed ones. It makes the locally unzipped SDK/runtime scenario very confusing.

FWIW, I have `DOTNET_MULTILEVEL_LOOKUP=0` set in the global environment on my Windows machine today. Multilevel-lookup never does what I want, but I do realize that I may be special (running preview builds by just unziping them, etc.).

Reply:
So do you think we should leave the current behavior where on Windows it does something but on Linux/macOS it's effectively ignored?

Reply:
I think that the problem with multi-level lookup is that it mixes local unzipped installs with global ones. If it stopped doing that, it would be fine with me (on both Windows and non-Windows).

Would something like this work (on both Windows and non-Windows)?

- For `dotnet` run from official global install, stitch together all global locations. (We would need a bread crumb that indentifies official global install.)
- For `dotnet` run from local unzipped install, look in the directory where `dotnet` resides or `DOTNET_ROOT` points to only.

Reply:
Just curious what is the scenario which breaks if the private install is mixed with the global ones?
Is it cases where you want an app to only run using the private install bits even if there's a better match in the global ones?

Your proposal would probably work - in almost all cases there will only be one global location, so it's kind of a moot point for the `dotnet` from global location. (Given current installers, the only way to get two global locations is to install 3.0 into custom location and then install 2.0 which always goes to the default location).

Basically what you're asking for (almost) is to disable multi-level lookup by default. Which is a breaking change. There's a problem with where the code lives. The framework resolution code is in `hostfxr` and so this goes back to the discussion below of how `dotnet` should search for `hostfxr`. If we keep the current behavior (only look next to itself), then the change is probably not really breaking, although it would mean 3.0 behaves differently from 2.0. If we allow `dotnet` to load higher version `hostfxr` we would have to somehow change the behavior of `hostfxr` based on the version of `dotnet` which called it.

The other consideration is `apphost` (and the other hosts). Per your proposal what should they base the decision on? The location of `hostfxr` which they end up using (from global location or private one)?

There's also another angle: dotnet/core-setup#4915. Which asks for `DOTNET_ROOT` to have effect on globally installed `dotnet`. It would effectively allow the reverse direction: using global `dotnet` with private install of shared frameworks and so on.

Reply:
> Just curious what is the scenario which breaks if the private install is mixed with the global ones?

Unzip a particular build (preview, servicing, ...) and do something against it. You may or may not be running the private unzipped bits with multi-level lookup depending on the global runtimes you happened to have installed.

> The other consideration is apphost (and the other hosts).

You have to have `DOTNET_ROOT` set for apphosts to pick up your local unzipped bits independent on multi-level lookup. I would make `DOTNET_ROOT` to be exclusive and not stitch it together with the global locations.


Do we know what was the motivating scenario for which we have introduced the multi-level lookup in the first place, and why it was done for Windows only? What is the name of the customer that will be unhappy if we disable multi-level lookup?

One of the customers for private unzipped bits is our own tooling. We are disabling the multi-level lookup wherever we can: https://github.com/dotnet/arcade/search?q=DOTNET_MULTILEVEL_LOOKUP&unscoped_q=DOTNET_MULTILEVEL_LOOKUP . And when we forget to disable it somewhere, it always hunts us back (e.g. https://github.com/dotnet/coreclr/issues/22971). I think our own tooling is going to be in strictly happier place if the multi-level lookup is disabled by default.

Reply:
Per this: https://github.com/dotnet/core-setup/issues/3606#issuecomment-363164964 a big user is ASP.NET.

Reply:
Do they still need it after move to Arcade?

Reply:
I spot checked aspnet/AspNetCore repo and it disables it just like all the other repos... interesting. I mean it makes perfect sense for repos to isolate themselves, for predictability.
I don't know why we introduced it in the first place... I will try to find who did it and figure out the reasoning. Unfortunately this a well document feature even externally. And also unfortunately it's on by default, so users may rely on it without even knowing about it.

Reply:
> I don't know why we introduced it in the first place.

I suspect it was to make the original `dnx` & `project.json` workflows work that are history by now

Reply:
The multi-level lookup for frameworks was introduced in dotnet/core-setup#1241 and the SDK counterpart then in dotnet/core-setup#1568. It has been introduced first in .NET Core 2.0.
Unfortunately it seems that most people involved in the original work have left... I guess the best bet is @eerhardt and @nguerrera .

Short version: We're looking for thoughts on basically changing the `DOTNET_MULTILEVEL_LOOKUP` to off by default (possibly with some minor tweaks to the exact behavior).

Reply:
I'll try to summarize some history as I remember it...

When first implemented, multi-level lookup search several more levels. It looks in some well known offset from CWD and in a well known offset from user profile. When I learned about this, I reached out questioning the wisdom of CWD, and after several discussions, it was determined that both CWD and user profile should be removed. This left x-plat with no other locations to search. At the time, there was no apphost and no code that tried a hard-coded location in /usr/ as a global location x-plat. 

The user profile scenario case was intended for this workflow that we have yet to deliver as the convenience of the experience never seems to get past other concerns including security:

1. Repo build/init script pulls down version repo needs to user profile (dotnet-install even defaults to this location IIRC)
2. Global.json in repo points to (1)
3. From there, double clicking on .sln in repo makes VS use (1) without setting any environment variables prior to VS launch

Essentially, this allows a zero admin install / zero global configuration: just clone, build, launch in IDE however you like. 

User profile was used in place of a private location inside the repo so that repos could share. Once you built one repo with that version, you didn't need to re-download and waste space. This was particularly interesting in a world of 50 ASP.NET repos. 

I believe that ASP.NET ended up using the closest equivalent to this under the degenerate multi-level lookup (dotnet dir + global location) by setting PATH once on their machines to the user location. It has probably changed again since then.Later on, putting dotnet on PATH alone was not good enough with the appearance of apphosts and DOTNET_ROOT. You now have to set both PATH and DOTNET_ROOT to have a full experience in the IDE.

I still think we're missing this scenario, and I do think it would be nice to go after some variant of it. Environment variables as per-repo configuration are really painful. We get a ton of negative feedback about how hard it is to use a private install in the IDE. No matter how many instructions are put in the repo docs, they are simply not read. Anything other than clone/build/open in VS leads to upset contributors. 

We have informally talked about the alternate possibility of putting a PATH to dotnet location in global.json, which repo build would arrange to download to. For this to actually work as a replacement for the original vision, it would have to be respected as DOTNET_ROOT. I suspect we're going to have about the same difficulty there as with the original multi-level lookup vision. :(

cc @davidfowl

Reply:
I agree that multi-level lookup interaction "unzip and use directly from the unzipped location" is extremely surprising behavior. I have spent a ton of time helping people getting tripped up by it. I also turn off multi-level lookup globally on my machines. Where we are now, we are getting leftover surprises from a feature that ended up being crippled to the point of not providing its original intent.

I also agree that searching multiple global, privileged, registered locations should not be something we need to disable by default.